File: src/bosh-monitor/lib/bosh/monitor/events/heartbeat.rb
module Bosh::Monitor
  module Events
    class Heartbeat < Base
      attr_reader :agent_id, :deployment, :job, :index, :metrics, :instance_id, :teams

      def initialize(attributes = {})
        super
        @kind = :heartbeat
        @metrics = []

        @id = @attributes['id']
        @timestamp = begin
                       Time.at(@attributes['timestamp'])
                     rescue StandardError
                       @attributes['timestamp']
                     end

        @deployment = @attributes['deployment']
        @agent_id = @attributes['agent_id']
        @job = @attributes['job']
        @index = @attributes['index'].to_s
        @instance_id = @attributes['instance_id']
        @job_state = @attributes['job_state']
        @teams = @attributes['teams']

        @tags = {}
        @tags['job'] = @job if @job
        @tags['index'] = @index if @index
        @tags['id'] = @instance_id if @instance_id

        @vitals = @attributes['vitals'] || {}
        @load = @vitals['load'] || []
        @cpu = @vitals['cpu'] || {}
        @mem = @vitals['mem'] || {}
        @swap = @vitals['swap'] || {}
        @disk = @vitals['disk'] || {}
        @system_disk = @disk['system'] || {}
        @ephemeral_disk = @disk['ephemeral'] || {}
        @persistent_disk = @disk['persistent'] || {}

        populate_metrics
      end

      def validate
        add_error('id is missing') if @id.nil?
        add_error('timestamp is missing') if @timestamp.nil?

        add_error('timestamp is invalid') if @timestamp && !@timestamp.is_a?(Time)
      end

      def add_metric(name, value)
        @metrics << Metric.new(name, value, @timestamp.to_i, @tags) if value
      end

      def short_description
        description = "Heartbeat from #{@job}/#{@instance_id} (agent_id=#{@agent_id}"

        description += " index=#{@index}" if @index && !@index.empty?

        description + ") @ #{@timestamp.utc}"
      end

      def to_s
        short_description
      end

      def to_hash
        result = {
          kind: 'heartbeat',
          id: @id,
          timestamp: @timestamp.to_i,
          deployment: @deployment,
          agent_id: @agent_id,
          job: @job,
          index: @index,
          instance_id: @instance_id,
          job_state: @job_state,
          vitals: @vitals,
          teams: @teams,
          metrics: @metrics.map(&:to_hash),
        }
        # Include process_length if present in attributes
        result[:process_length] = @attributes['process_length'] if @attributes.key?('process_length')
        result
      end

      def to_json(*_args)
        JSON.dump(to_hash)
      end

      def to_plain_text
        short_description
      end

      private

      def populate_metrics
        add_metric('system.load.1m', @load[0]) if @load.is_a?(Array)
        add_metric('system.cpu.user', @cpu['user'])
        add_metric('system.cpu.sys', @cpu['sys'])
        add_metric('system.cpu.wait', @cpu['wait'])
        add_metric('system.mem.percent', @mem['percent'])
        add_metric('system.mem.kb', @mem['kb'])
        add_metric('system.swap.percent', @swap['percent'])
        add_metric('system.swap.kb', @swap['kb'])
        add_metric('system.disk.system.percent', @system_disk['percent'])
        add_metric('system.disk.system.inode_percent', @system_disk['inode_percent'])
        add_metric('system.disk.ephemeral.percent', @ephemeral_disk['percent'])
        add_metric('system.disk.ephemeral.inode_percent', @ephemeral_disk['inode_percent'])
        add_metric('system.disk.persistent.percent', @persistent_disk['percent'])
        add_metric('system.disk.persistent.inode_percent', @persistent_disk['inode_percent'])
        add_metric('system.healthy', @job_state == 'running' ? 1 : 0)
      end
    end
  end
end

File: src/bosh-monitor/pkg/events/heartbeat.go
package events

import (
	"encoding/json"
	"fmt"
	"time"
)

type Heartbeat struct {
	HeartbeatID string
	Timestamp   time.Time
	Deployment  string
	AgentID     string
	Job         string
	Index       string
	InstanceID  string
	JobState    string
	Teams       []string
	Vitals      map[string]interface{}
	HBMetrics   []Metric
	Attrs       map[string]interface{}
}

func NewHeartbeat(attributes map[string]interface{}) *Heartbeat {
	h := &Heartbeat{Attrs: attributes}

	if v, ok := attributes["id"]; ok {
		h.HeartbeatID = fmt.Sprintf("%v", v)
	}
	if v, ok := attributes["timestamp"]; ok {
		switch tv := v.(type) {
		case int:
			h.Timestamp = time.Unix(int64(tv), 0)
		case int64:
			h.Timestamp = time.Unix(tv, 0)
		case float64:
			h.Timestamp = time.Unix(int64(tv), 0)
		case time.Time:
			h.Timestamp = tv
		}
	}
	if v, ok := attributes["deployment"]; ok {
		h.Deployment = fmt.Sprintf("%v", v)
	}
	if v, ok := attributes["agent_id"]; ok {
		h.AgentID = fmt.Sprintf("%v", v)
	}
	if v, ok := attributes["job"]; ok {
		h.Job = fmt.Sprintf("%v", v)
	}
	if v, ok := attributes["index"]; ok {
		h.Index = fmt.Sprintf("%v", v)
	}
	if v, ok := attributes["instance_id"]; ok {
		h.InstanceID = fmt.Sprintf("%v", v)
	}
	if v, ok := attributes["job_state"]; ok {
		h.JobState = fmt.Sprintf("%v", v)
	}
	if v, ok := attributes["teams"]; ok {
		if teams, ok := v.([]interface{}); ok {
			for _, t := range teams {
				h.Teams = append(h.Teams, fmt.Sprintf("%v", t))
			}
		} else if teams, ok := v.([]string); ok {
			h.Teams = teams
		}
	}
	if v, ok := attributes["vitals"]; ok {
		if vitals, ok := v.(map[string]interface{}); ok {
			h.Vitals = vitals
		}
	}
	if h.Vitals == nil {
		h.Vitals = map[string]interface{}{}
	}

	h.populateMetrics()
	return h
}

func (h *Heartbeat) ID() string   { return h.HeartbeatID }
func (h *Heartbeat) Kind() string { return "heartbeat" }

func (h *Heartbeat) Validate() []string {
	var errs []string
	if h.HeartbeatID == "" {
		errs = append(errs, "id is missing")
	}
	if h.Timestamp.IsZero() {
		errs = append(errs, "timestamp is missing")
	}
	return errs
}

func (h *Heartbeat) Valid() bool {
	return len(h.Validate()) == 0
}

func (h *Heartbeat) ShortDescription() string {
	desc := fmt.Sprintf("Heartbeat from %s/%s (agent_id=%s", h.Job, h.InstanceID, h.AgentID)
	if h.Index != "" {
		desc += fmt.Sprintf(" index=%s", h.Index)
	}
	desc += fmt.Sprintf(") @ %s", h.Timestamp.UTC().Format(time.RFC1123Z))
	return desc
}

func (h *Heartbeat) ToHash() map[string]interface{} {
	result := map[string]interface{}{
		"kind":        "heartbeat",
		"id":          h.HeartbeatID,
		"timestamp":   h.Timestamp.Unix(),
		"deployment":  h.Deployment,
		"agent_id":    h.AgentID,
		"job":         h.Job,
		"index":       h.Index,
		"instance_id": h.InstanceID,
		"job_state":   h.JobState,
		"vitals":      h.Vitals,
		"teams":       h.Teams,
		"metrics":     h.metricsToHash(),
	}
	if v, ok := h.Attrs["number_of_processes"]; ok {
		result["number_of_processes"] = v
	}
	return result
}

func (h *Heartbeat) metricsToHash() []map[string]interface{} {
	var result []map[string]interface{}
	for _, m := range h.HBMetrics {
		result = append(result, map[string]interface{}{
			"name":      m.Name,
			"value":     m.Value,
			"timestamp": m.Timestamp,
			"tags":      m.Tags,
		})
	}
	return result
}

func (h *Heartbeat) ToJSON() (string, error) {
	data, err := json.Marshal(h.ToHash())
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func (h *Heartbeat) ToPlainText() string {
	return h.ShortDescription()
}

func (h *Heartbeat) Metrics() []Metric {
	return h.HBMetrics
}

func (h *Heartbeat) Attributes() map[string]interface{} {
	return h.Attrs
}

func (h *Heartbeat) String() string {
	return h.ShortDescription()
}

func (h *Heartbeat) addMetric(name string, value interface{}) {
	if value == nil {
		return
	}
	tags := map[string]string{}
	if h.Job != "" {
		tags["job"] = h.Job
	}
	if h.Index != "" {
		tags["index"] = h.Index
	}
	if h.InstanceID != "" {
		tags["id"] = h.InstanceID
	}
	h.HBMetrics = append(h.HBMetrics, Metric{
		Name:      name,
		Value:     fmt.Sprintf("%v", value),
		Timestamp: h.Timestamp.Unix(),
		Tags:      tags,
	})
}

func (h *Heartbeat) populateMetrics() {
	load := getSlice(h.Vitals, "load")
	if len(load) > 0 {
		h.addMetric("system.load.1m", load[0])
	}

	cpu := getMap(h.Vitals, "cpu")
	h.addMetric("system.cpu.user", cpu["user"])
	h.addMetric("system.cpu.sys", cpu["sys"])
	h.addMetric("system.cpu.wait", cpu["wait"])

	mem := getMap(h.Vitals, "mem")
	h.addMetric("system.mem.percent", mem["percent"])
	h.addMetric("system.mem.kb", mem["kb"])

	swap := getMap(h.Vitals, "swap")
	h.addMetric("system.swap.percent", swap["percent"])
	h.addMetric("system.swap.kb", swap["kb"])

	disk := getMap(h.Vitals, "disk")
	systemDisk := getMap(disk, "system")
	h.addMetric("system.disk.system.percent", systemDisk["percent"])
	h.addMetric("system.disk.system.inode_percent", systemDisk["inode_percent"])

	ephemeralDisk := getMap(disk, "ephemeral")
	h.addMetric("system.disk.ephemeral.percent", ephemeralDisk["percent"])
	h.addMetric("system.disk.ephemeral.inode_percent", ephemeralDisk["inode_percent"])

	persistentDisk := getMap(disk, "persistent")
	h.addMetric("system.disk.persistent.percent", persistentDisk["percent"])
	h.addMetric("system.disk.persistent.inode_percent", persistentDisk["inode_percent"])

	healthy := 0
	if h.JobState == "running" {
		healthy = 1
	}
	h.addMetric("system.healthy", healthy)
}

func getMap(m map[string]interface{}, key string) map[string]interface{} {
	if v, ok := m[key]; ok {
		if vm, ok := v.(map[string]interface{}); ok {
			return vm
		}
	}
	return map[string]interface{}{}
}

func getSlice(m map[string]interface{}, key string) []interface{} {
	if v, ok := m[key]; ok {
		if vs, ok := v.([]interface{}); ok {
			return vs
		}
	}
	return nil
}
